qiskit/primitives/containers/observables_array.py

```diff
@@ -70,31 +66,51 @@ def __init__(
                 the input should already be an array-like.
 
         Raises:
-            ValueError: If ``validate=True`` and the input observables is not valid.
+            ValueError: If ``validate=True`` and the input observables array is not valid.
         """
         super().__init__()
         if isinstance(observables, ObservablesArray):
             observables = observables._array
         self._array = object_array(observables, copy=copy, list_types=(PauliList,))
         self._shape = self._array.shape
         if validate:
-            num_qubits = None
             for ndi, obs in np.ndenumerate(self._array):
-                basis_obs = self.coerce_observable(obs)
-                basis_num_qubits = len(next(iter(basis_obs)))
-                if num_qubits is None:
-                    num_qubits = basis_num_qubits
-                elif basis_num_qubits != num_qubits:
-                    raise ValueError(
-                        "The number of qubits must be the same for all observables in the "
-                        "observables array."
-                    )
-                self._array[ndi] = basis_obs
+                self._array[ndi] = self.coerce_observable(obs)
+
+    @staticmethod
+    def obs_to_dict(obs: SparseObservable) -> Mapping[str, float]:
+        """Convert a sparse observable to a mapping from Pauli strings to coefficients"""
+        sparse_list = obs.to_sparse_list()
+        result = {}
+        for pauli_term in sparse_list:
+            sparse_pauli_str, pauli_qubits, coeff = pauli_term
+
+            if len(sparse_pauli_str) == 0:
+                full_pauli_str = "I" * obs.num_qubits
+            else:
+                sorted_lists = sorted(zip(pauli_qubits, sparse_pauli_str))
+                pauli_qubits = [pq for pq, _ in sorted_lists]
+                sparse_pauli_str = [spstr for _, spstr in sorted_lists]
+
+                full_pauli_str = ""
+                prev_qubit = -1
+                for qubit, pauli in zip(pauli_qubits, sparse_pauli_str):
+                    full_pauli_str += "I" * (qubit - prev_qubit - 1) + pauli
+                    prev_qubit = qubit
+
+                full_pauli_str += "I" * (obs.num_qubits - pauli_qubits[-1] - 1)
+                full_pauli_str = full_pauli_str[::-1]
```

Comment:
this implementation could be tidied up a bit: 
 - don't `+=` strings in loops
 - unzipping then immediately rezipping the same things

Reply:
> unzipping then immediately rezipping the same things

Please check e212dc0d7, if this is what you meant, in particular the usage of `max(pauli_qubits)` (linear) instead of `pauli_qubits[-1]` (constant).